src/SimpleRADOSStriper.cc

```diff
@@ -488,7 +488,7 @@ ssize_t SimpleRADOSStriper::read(void* data, size_t len, uint64_t off)
   }
 
   size_t r = 0;
-  std::vector<std::pair<bufferlist, aiocompletionptr>> reads;
+  std::deque<std::pair<bufferlist, aiocompletionptr>> reads;
```

Comment:
OK, this is the culprit:

```cpp

  while ((len-r) > 0) {
    // ...
    auto& [bl, aiocp] = reads.emplace_back();
    aiocp = aiocompletionptr(librados::Rados::aio_create_completion());
    if (int rc = ioctx.aio_read(ext.soid, aiocp.get(), &bl, ext.len, ext.off); rc < 0) {
```

Issuing a new read grows the vector, and thus may invalidate (if the vector needs to reallocate) the references used for previous reads. `std::deque` should do the trick in this use case. Still, worth a comment.

Reply:
One more thing. Let's document we need to be very careful with `std::deque` here. `emplace_back()` is fine as it's obliged to preserve the references but, for instance, `insert()` nor `erase()` don't offer such a guarantee.